app/controllers/application_controller.rb

```diff
@@ -5,6 +5,148 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "password_security"
+  end
+### INDEX ###
+  get '/' do 
+    erb :index
+  end
+
+### SIGNUP ###
+  get '/signup' do 
+    if Helpers.is_logged_in?(session)
+      redirect "/tweets"
+    else
+      erb :'/users/create_user'
+    end
+  end
+
+  post '/signup' do 
+    @user = User.new(username: params["username"], email: params["email"], password: params["password"])
+    if !@user.username.empty? && !@user.email.empty? && @user.password != nil
+      @user.save
+      session[:user_session_id] = @user.id
+      redirect "/tweets"
+    else
+      redirect "/signup"
+    end
+  end
+
+### LOGIN ###
+  get '/login' do
+    if Helpers.is_logged_in?(session)
+      redirect "/tweets"
+    else
+      erb :'/users/login'
+    end
+  end
+
+  post '/login' do 
+    @user = User.find_by(:username => params[:username])
+    if @user && @user.authenticate(params[:password])
+      session[:user_session_id] = @user.id
+      redirect "/tweets"
+    else
+      redirect "/login"
+    end
+  end
+
+### LOGOUT ###
+  get '/logout' do
+    if Helpers.is_logged_in?(session)
+      session.clear
+      redirect "/login"
+    else
+      redirect "/"
+    end
+  end
+
+  ### SHOW ALL TWEETS OF SPECIFIC USER ###
+  get '/users/:slug' do 
+    @user = User.find_by_slug(params[:slug])
+    erb :'users/show_tweets'
+  end
+
+
+### CREATE NEW TWEET ###
+  get '/tweets/new' do
+    if Helpers.is_logged_in?(session)
+      erb :'/tweets/create_tweet' 
+    else
+      redirect "/login"
+    end
+  end
+
+  post '/tweets/new' do 
+    @tweet = Tweet.new(:content => params["content"], :user_id => params["user_id"])
```

Comment:
The way I managed this (and I'm sure this is weird) is that in the view erb :'/tweets/create_tweet', I used erb to call Helpers.is_logged_in? as a conditional and if the user was logged in, I set a local variable "user" to Helpers.current_user. (so user=Helpers.current_user)

My concern was exactly what you mentioned, a user somehow posting a tweet as someone else, so I set the user_id param as a hidden input with a name of "user_id" and a value of <%=user.id%>, so that when I created the tweet in the controller's post '/tweets/new' route the tweet would belong to whomever was logged in.  Make sense?

It feels like a hidden input can still be messed with though, yeah?  Maybe by inspecting the element or something?  I honestly don't know but since it is being sent as an HTML request it's out there somehow and is less secure than taking care of it server side in the controller.

And I didn't even consider using the has_many relationship.  My brain just isn't there yet :) 

I don't think I even knew you could call create like that.  That's awesome.  


> On Apr 9, 2016, at 7:47 PM, Randall Reed <notifications@github.com> wrote:
> 
> In app/controllers/application_controller.rb <https://github.com/learn-co-students/sinatra-fwitter-group-project-v-000/pull/79#discussion_r59125769>:
> 
> > +    @user = User.find_by_slug(params[:slug])
> > +    erb :'users/show_tweets'
> > +  end
> > +
> > +
> > +### CREATE NEW TWEET ###
> > +  get '/tweets/new' do
> > +    if Helpers.is_logged_in?(session)
> > +      erb :'/tweets/create_tweet' 
> > +    else
> > +      redirect "/login"
> > +    end
> > +  end
> > +
> > +  post '/tweets/new' do 
> > +    @tweet = Tweet.new(:content => params["content"], :user_id => params["user_id"])
> user_id should not be in the params. A malicious user could use that to spoof another user's account and pose as them. Instead, use the session's user, current_user.
> 
> Another way to create the tweet and associate it to the user is to make use of the has_many relationship on User:
> 
> @tweet = current_user.tweets.create(content: params[:content])
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub <https://github.com/learn-co-students/sinatra-fwitter-group-project-v-000/pull/79/files/9dfc20de18dce246a6925028c350c9316e9e3937#r59125769>